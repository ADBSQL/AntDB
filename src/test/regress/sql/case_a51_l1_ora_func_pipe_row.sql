set grammar to oracle;

--pipe row


------数值类型
-------------------------------------
\set PLSQL_MODE on
CREATE OR REPLACE FUNCTION f1_t4test(x NUMBER) 
RETURN NUMBER PIPELINED 
IS
  BEGIN
    FOR i IN 1..x LOOP
      PIPE ROW(i);
    END LOOP;
    RETURN;
  END f1_t4test;
/

\set PLSQL_MODE off

select * from table(f1_t4test(5));

drop FUNCTION f1_t4test(x NUMBER);

----------字符类型+双行+ROWNUM
-----------------------------------
\set PLSQL_MODE on
CREATE OR REPLACE FUNCTION f2_t4test(x NUMBER) 
RETURN varchar2 PIPELINED 
IS
  BEGIN
    FOR i IN 1..x LOOP
      PIPE ROW('row1:'||to_char(i));
	  PIPE ROW('row2:'||to_char(i));
    END LOOP;
    RETURN;
  END f2_t4test;
/

\set PLSQL_MODE off

select * from table(f2_t4test(5)) where ROWNUM<6;

drop FUNCTION f2_t4test(x NUMBER);

---------------表达式
----------------------------------------------
\set PLSQL_MODE on
CREATE OR REPLACE FUNCTION f3_t4test(x NUMBER) 
RETURN varchar2 PIPELINED 
IS
  BEGIN
    FOR i IN 1..x LOOP
      PIPE ROW(((i+2)*2)/i);
    END LOOP;
    RETURN;
  END f3_t4test;
/

\set PLSQL_MODE off

select * from table(f3_t4test(5));

drop FUNCTION f3_t4test(x NUMBER);
---------------存储过程
----------------------------------------------
\set PLSQL_MODE on

create or replace procedure t4test_pro(id in int,name1 out VARCHAR2, name2 out VARCHAR2)
as  
        begin
          name1:='name1:'||to_char(id);
		  name2:='name2:'||to_char(id);
         end;
/


CREATE OR REPLACE FUNCTION f4_t4test(x NUMBER) 
RETURN varchar2 PIPELINED 
IS
  BEGIN
    FOR i IN 1..x LOOP
      PIPE ROW(select t4test_pro(i));
    END LOOP;
    
  END f4_t4test;
/

\set PLSQL_MODE off

select * from table(f4_t4test(5));

drop FUNCTION f4_t4test(x NUMBER);
drop procedure t4test_pro(id in int,name1 out VARCHAR2, name2 out VARCHAR2);








