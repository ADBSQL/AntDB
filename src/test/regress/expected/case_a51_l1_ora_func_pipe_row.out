set grammar to oracle;
--pipe row
------数值类型
-------------------------------------
\set PLSQL_MODE on
CREATE OR REPLACE FUNCTION f1_t4test(x NUMBER) 
RETURN NUMBER PIPELINED 
IS
  BEGIN
    FOR i IN 1..x LOOP
      PIPE ROW(i);
    END LOOP;
    RETURN;
  END f1_t4test;
/
\set PLSQL_MODE off
select * from table(f1_t4test(5));
 COLUMN_VALUE 
--------------
            1
            2
            3
            4
            5
(5 rows)

drop FUNCTION f1_t4test(x NUMBER);
----------字符类型+双行+ROWNUM
-----------------------------------
\set PLSQL_MODE on
CREATE OR REPLACE FUNCTION f2_t4test(x NUMBER) 
RETURN varchar2 PIPELINED 
IS
  BEGIN
    FOR i IN 1..x LOOP
      PIPE ROW('row1:'||to_char(i));
	  PIPE ROW('row2:'||to_char(i));
    END LOOP;
    RETURN;
  END f2_t4test;
/
\set PLSQL_MODE off
select * from table(f2_t4test(5)) where ROWNUM<6;
 COLUMN_VALUE 
--------------
 row1:1
 row2:1
 row1:2
 row2:2
 row1:3
(5 rows)

drop FUNCTION f2_t4test(x NUMBER);
---------------表达式
----------------------------------------------
\set PLSQL_MODE on
CREATE OR REPLACE FUNCTION f3_t4test(x NUMBER) 
RETURN varchar2 PIPELINED 
IS
  BEGIN
    FOR i IN 1..x LOOP
      PIPE ROW(((i+2)*2)/i);
    END LOOP;
    RETURN;
  END f3_t4test;
/
\set PLSQL_MODE off
select * from table(f3_t4test(5));
    COLUMN_VALUE    
--------------------
 6
 4
 3.3333333333333333
 3
 2.8
(5 rows)

drop FUNCTION f3_t4test(x NUMBER);
---------------存储过程
----------------------------------------------
\set PLSQL_MODE on
create or replace procedure t4test_pro(id in int,name1 out VARCHAR2, name2 out VARCHAR2)
as  
        begin
          name1:='name1:'||to_char(id);
		  name2:='name2:'||to_char(id);
         end;
/
CREATE OR REPLACE FUNCTION f4_t4test(x NUMBER) 
RETURN varchar2 PIPELINED 
IS
  BEGIN
    FOR i IN 1..x LOOP
      PIPE ROW(select t4test_pro(i));
    END LOOP;
    
  END f4_t4test;
/
\set PLSQL_MODE off
select * from table(f4_t4test(5));
   COLUMN_VALUE    
-------------------
 (name1:1,name2:1)
 (name1:2,name2:2)
 (name1:3,name2:3)
 (name1:4,name2:4)
 (name1:5,name2:5)
(5 rows)

drop FUNCTION f4_t4test(x NUMBER);
drop procedure t4test_pro(id in int,name1 out VARCHAR2, name2 out VARCHAR2);
